pennylane/devices/default_mixed.py

```diff
@@ -449,6 +452,50 @@ def _apply_density_matrix(self, state, device_wires):
             )
             self._state = self._asarray(rho, dtype=self.C_DTYPE)
 
+    def _generate_projector_diagonal(self, wire):
+        n_wires_right = self.num_wires-wire-1
+        n_wires_left = wire
+        id_dim_right = 2**n_wires_right
+        id_dim_left = 2**n_wires_left
+
+        diag = [1]*(id_dim_right) + [0]*(id_dim_right)
+        diag = diag*(id_dim_left)
+
+        return diag
+
+    def _apply_measurement(self, meas_op):
+        zero_projector = self._generate_projector_diagonal(meas_op.wire)
+        one_projector = [int(not b) for b in zero_projector]
+
+        pre_op_state = self._state
+        self._apply_diagonal_unitary(zero_projector, self.wires)
+        zero_state = self.state
+
+        self._state = pre_op_state
+        self._apply_diagonal_unitary(one_projector, self.wires)
+        one_state = self.state
+
+        rho = zero_state + one_state
+        self._apply_density_matrix(rho, self.wires)
+
+    def _apply_conditional(self, if_op):
+        zero_projector = self._generate_projector_diagonal(if_op.m_val.wire)
+        one_projector = [int(not b) for b in zero_projector]
+
+        pre_op_state = self._state
+        self._apply_diagonal_unitary(one_projector, self.wires)
+        self._apply_operation(if_op.then_op)
+        one_state = self.state
+
+        self._state = pre_op_state
+        self._apply_diagonal_unitary(zero_projector, self.wires)
+        if if_op.else_op is not None:
+            self._apply_operation(if_op.else_op)
+        zero_state = self.state
+
+        rho = zero_state + one_state
+        self._apply_density_matrix(rho, self.wires)
```

Comment:
Before applying the dm, don't we have to set the state back? I.e.,
```python
    self._state = pre_op_state
```

Reply:
I don't think so because `_apply_density_matrix` sets the internal state so it doesn't matter what it was before.